src/coreclr/tools/dotnet-pgo/Program.cs

```diff
@@ -1382,8 +1390,75 @@ private int InnerProcessTraceFileMain()
                             continue;
                         }
 
-                        if ((e.TimeStampRelativeMSec >= excludeEventsBefore) && (e.TimeStampRelativeMSec <= excludeEventsAfter))
-                            methodsToAttemptToPrepare.Add((int)e.EventIndex, new ProcessedMethodData(e.TimeStampRelativeMSec, method, "R2RLoad"));
+                        if (e.TimeStampRelativeMSec < excludeEventsBefore)
+                        {
+                            continue;
+                        }
+
+                        if (e.TimeStampRelativeMSec > excludeEventsAfter)
+                        {
+                            break;
+                        }
+
+                        if (includeMethods != null || excludeMethods != null)
+                        {
+                            string methodAsString = method.ToString();
```

Comment:
This string is generally meant for debugging, it goes through `DebugNameFormatter` which will produce method names that users may not be familiar with (e.g. I believe types will be module prefix in the format "[module name]type name"). Would it be better to introduce a custom formatter, reuse the `TypeString` formatter, or to reuse the `methodNameFromEventDirectly` that we are already using to print the warnings above?

Reply:
Is there expectation for users to be more specific than to use a regular expression for specifying which method to exclude events after? I was trying to get a local test loop going on Android to see the structural differences between a `MethodDesc.ToString()` and the `methodNameFromEventDirectly` but its taking me some time to get it working locally.

I can switch it to `methodNameFromEventDirectly` as that should still be able to match a regex provided by the user. Is it usually easier to match the `methodNameFromEventDirectly` form than to match the `MethodDesc.ToString()` form even though it is passed through the `DebugNameFormatter`?

Reply:
The problem is that the way `DebugNameFormatter` formats names is an internal detail of the managed type system, so I do not think user-facing tools should be using this for any purpose but diagnostics. We do not want situations where we change this detail in the future and break something that took a dependency on the exact way that namespaces or class names are formatted by this formatter.

I looked again and realize `methodNameFromEventDirectly` is not actually a field that comes directly from the event, but a concatenation of various fields. I think a simpler concatenation like `e.MethodNamespace + "." + e.MethodName` would be better to do the matching against. I think `MethodNamespace` includes the class name, but you may want to double check.

Reply:
I think this also suggests that `e.MethodNamespace` will also contain the class name https://github.com/microsoft/perfview/blob/766a08ca8373d273d55da8b07147e7068e4d5f58/src/TraceEvent/Samples/21_ObserveJitEvents.cs#L126-L132.

Maybe `methodNameFromEventDirectly` would allow for more flexibility as it will allow users to also specify between different overloaded forms of the same method? It would also more closely align with a `GetMethodName` in Perfview https://github.com/microsoft/perfview/blob/766a08ca8373d273d55da8b07147e7068e4d5f58/src/TraceEvent/Computers/TraceManagedProcess.cs#L4170

Reply:
> Do you know where exactly this name is displayed in the PerfView UI?

No, my main dev machine is osx which I think has problems with perfview (or atleast it did a year or two ago when I tried), I'd have to sleuth through to find out or try to setup something on my windows laptop.

> We should also update documentation of the CLI option

I can add an update to the description `src/coreclr/tools/dotnet-pgo/PgoRootCommand.cs`, thats what you're referring to right? Or is there another file for documentation that this PR currently doesn't touch?

> we should match the PerfView name exactly

I can create another var `perfviewMethodName` so it'll be clearer that we are matching the same pattern as `Microsoft.Diagnostics.Tracing.Analysis.JIT.JITStats.GetMethodName`:  `$"{e.MethodNamespace}.{e.MethodName}{paramsArgs}"`

Reply:
> I would also add a note that this includes parameters, so that users can realize they cannot use e.g. *MyClass* to match all methods defined on that class

Oh right, as it would also match other methods that have `MyClass` as a parameter? Then would the only way to match all methods in `*MyClass*` be to have a regex expression of `^MyNamespace\.MyClass.*`

Reply:
> No, my main dev machine is osx which I think has problems with perfview (or atleast it did a year or two ago when I tried), I'd have to sleuth through to find out or try to setup something on my windows laptop.
> I can create another var perfviewMethodName so it'll be clearer that we are matching the same pattern as Microsoft.Diagnostics.Tracing.Analysis.JIT.JITStats.GetMethodName: $"{e.MethodNamespace}.{e.MethodName}{paramsArgs}"

Ok, so it's the name from the JITStats view within PerfView. Sounds good to add a separate `perfviewMethodName` for it. I took a look and here are some examples of what that looks like:

* ``System.Collections.Generic.Dictionary`2[Microsoft.VisualStudio.Telemetry.WindowsMachinePropertyProvider+ProcessorArchitectureType,System.__Canon].get_Item(!0)``
* ``System.Lazy`1[Microsoft.VisualStudio.Telemetry.WindowsOSPropertyProvider+DisplayInformation]..ctor(class System.Func`1,value class System.Threading.LazyThreadSafetyMode)``
* ``Newtonsoft.Json.Serialization.JsonSerializerInternalReader.EndProcessProperty(class System.Object,class Newtonsoft.Json.JsonReader,class Newtonsoft.Json.Serialization.JsonObjectContract,int32,class Newtonsoft.Json.Serialization.JsonProperty,value class PropertyPresence,bool``
* ``System.Linq.Enumerable.ToDictionary(class System.Collections.Generic.IEnumerable`1,class System.Func`2,class System.Func`2,class System.Collections.Generic.IEqualityComparer`1)`` (so apparently method instantiations are not included)

This is for coreclr, I'm not sure how this is going to look for Mono. FWIW, we also have #87257 which is going to change this format (in particular the `!0` in the `get_Item` example would be replaced by the instantiation, I believe).

> I can add an update to the description `src/coreclr/tools/dotnet-pgo/PgoRootCommand.cs`, thats what you're referring to right?

Right, exactly.

> Oh right, as it would also match other methods that have `MyClass` as a parameter? Then would the only way to match all methods in `*MyClass*` be to have a regex expression of `^MyNamespace\.MyClass.*`

Yep. I'm not sure if it would be a problem, or if we'd rather just have a simpler format that only includes the metadata class names and metadata method names.

Reply:
I'm not too familiar with the user scenarios, so I dont know if it is likely that one would want to only include methods with a certain set of parameters and not include the others. Maybe we can start out with having the most customizable option first (allowing users to include/exclude particular overloaded methods) and later if we know users will never want to do that we can pursue the simpler format?

Reply:
I'm also noticing that the `excludeEventsBeforeJittingMethod` and `excludeEventsAfterJittingMethod` are based off of `method.ToString()`. It would probably be better to maintain parity for those as well. And on more deliberation and offline discussion, maybe keeping the simpler format is better for now. Also, do you happen to know if `MethodDesc.ToString()` will contain the parameter information or not @jakobbotsch ?

Reply:
Ah, it looks like it does https://github.com/dotnet/runtime/blob/f5889ec694ebecdb8688ce3e1b987569caf41348/src/coreclr/tools/Common/TypeSystem/Common/MethodDesc.ToString.cs#L70-L79

Maybe it would be better to stick with @lateralusX's original intent of the regex matching.

Reply:
> I'm also noticing that the `excludeEventsBeforeJittingMethod` and `excludeEventsAfterJittingMethod` are based off of `method.ToString()`. It would probably be better to maintain parity for those as well. 

Yes, the code above should be changed in the same way. 

We should not use `MethodDesc.ToString()` for anything except diagnostics when it is user facing like this. It should have a dedicated formatter then if we do not want to use the name from the event. I think you can base it on [TypeString](https://github.com/dotnet/runtime/blob/main/src/coreclr/tools/aot/ILCompiler.ReadyToRun/ObjectWriter/TypeString.cs), except without including instantiations.

> I'm not very familiar with these types, would `e.MethodNamespace + "." + e.MethodName + paramsArgs` be the same values that are held by the resulting `MethodDesc` discovered by `idParser.ResolveMethoID(e.MethodID...)`? If not, then it might not be the right "substitution"

Yes, they represent the same thing. dotnet-pgo has a managed implementation of the type system that replicates the native implementation of the type system that produced the strings in the event. Classes and methods are loaded into the type system based on events in the trace.

Reply:
Updated to use `e.MethodNamespace + "." + e.MethodName + paramsArgs` as the method's name for regex matching